### PR TITLE
Adds support for treating missing headers as empty

### PR DESCRIPTION
**Pull request by davinci26**
`TreatMissingHeadersAsEmpty` specifies if the header match rule specified header does not exist, this header value will be treated as empty. Defaults to false. Unlike the underlying Envoy implementation this is **only** supported for negative matches (e.g. NotContains, NotExact).

The reason that I implemented only for negative matches is that is substantially simpler to reason about. I am open to implementing it for all cases if you think it is going to be useful.

### CODE REVIEW

**Comment by davinci26** [comment]
@clayton-gonsalves wanna do the first pass?

File: internal/featuretests/v3/httpproxy.go [hunk 1]
```diff
@@ -39,12 +39,13 @@ func headerContainsMatchCondition(name, value string, ignoreCase bool) contour_a
...
+func headerNotContainsMatchCondition(name, value string, ignoreCase, treatMissingAsEmpty bool) contour_api_v1.MatchCondition {
 	return contour_api_v1.MatchCondition{
 		Header: &contour_api_v1.HeaderMatchCondition{
-			Name:        name,
-			NotContains: value,
-			IgnoreCase:  ignoreCase,
+			Name:                       name,
+			NotContains:                value,
+			IgnoreCase:                 ignoreCase,
+			TreatMissingHeadersAsEmpty: treatMissingAsEmpty,
```
Comment (clayton-gonsalves) [inline internal/featuretests/v3/httpproxy.go:48]: Nit: I think this variable can be renamed as `TreatMissingAsEmpty` to maintain uniformity with other fields.

**Review by clayton-gonsalves — CHANGES_REQUESTED** [review]

File: internal/dag/dag.go [hunk 2]
```diff
@@ -140,18 +140,20 @@ const (
...
+	IgnoreCase                 bool
+	TreatMissingHeadersAsEmpty bool
 }
 
 func (hc *HeaderMatchCondition) String() string {
 	details := strings.Join([]string{
 		"name=" + hc.Name,
 		"value=" + hc.Value,
 		"matchtype=", hc.MatchType,
+		"TreatMissingHeadersAsEmpty=", strconv.FormatBool(hc.TreatMissingHeadersAsEmpty),
```
[the remark this answers is not in the document]
Comment (davinci26) [reply internal/dag/dag.go]: fixed in eb4feb66d4b58b05572e6230440a08b76cb1f6e0

File: internal/featuretests/v3/headercondition_test.go [hunk 3]
```diff
@@ -55,25 +55,38 @@ func TestConditions_ContainsHeader_HTTProxy(t *testing.T) {
...
+				{
+					Conditions: matchconditions(
+						prefixMatchCondition("/blog"),
+						headerNotExactMatchCondition("x-beta-release", "true", false, true),
+					),
+					Services: []contour_api_v1.Service{{
+						Name: "svc2",
+						Port: 80,
+					}},
+				},
```
[the remark this answers is not in the document]
Comment (davinci26) [reply internal/featuretests/v3/headercondition_test.go:88]: fixed in eb4feb66d4b58b05572e6230440a08b76cb1f6e0

File: Makefile [hunk 4]
```diff
@@ -19,7 +19,7 @@ CONTOUR_E2E_LOCAL_HOST ?= $(LOCALIP)
 # Variables needed for running e2e and upgrade tests.
 CONTOUR_UPGRADE_FROM_VERSION ?= $(shell ./test/scripts/get-contour-upgrade-from-version.sh)
 CONTOUR_E2E_IMAGE ?= ghcr.io/projectcontour/contour:main
-CONTOUR_E2E_PACKAGE_FOCUS ?= ./test/e2e
+CONTOUR_E2E_PACKAGE_FOCUS ?= ./test/e2e/httpproxy
```
[the remark this answers is not in the document]
Comment (davinci26) [reply Makefile]: fixed in 61240ddd16885f9c15d7e12b80ea1523c0b84e8e

**Comment by davinci26** [comment]
broke myself with the other PR 👀

File: apis/projectcontour/v1/httpproxy.go [hunk 5]
```diff
@@ -137,6 +139,13 @@ type HeaderMatchCondition struct {
 	// value.
 	// +optional
 	Regex string `json:"regex,omitempty"`
+
+	// TreatMissingAsEmpty specifies if the header match rule specified header
+	// does not exist, this header value will be treated as empty. Defaults to false.
+	// Unlike the underlying Envoy implementation this is **only** supported for
+	// negative matches (e.g. NotContains, NotExact).
+	// +optional
+	TreatMissingAsEmpty bool `json:"TreatMissingAsEmpty,omitempty"`
```
Comment (skriss) [inline apis/projectcontour/v1/httpproxy.go:148]:
```suggestion
	TreatMissingAsEmpty bool `json:"treatMissingAsEmpty,omitempty"`
```

File: site/content/docs/main/config/request-routing.md [hunk 6]
```diff
@@ -105,8 +105,13 @@ Regex conditions **must** start with a `/` if they are present.
 
 #### Header conditions
 
-For `header` conditions there is one required field, `name`, six operator fields: `present`, `notpresent`, `contains`, `notcontains`, `exact`, and `notexact` and a modifier `ignoreCase` which can be used together with all of the operator fields except `regex`, `notpresent` and `present`.
+For `header` conditions there is the following structure:
 
+1. one required field, `name`
+2. six operator fields: `present`, `notpresent`, `contains`, `notcontains`, `exact`, and `notexact`
+3. two optional modifiers: `ignoreCase` and `TreatMissingAsEmpty`
```
Comment (skriss) [inline site/content/docs/main/config/request-routing.md:112]:
```suggestion
3. two optional modifiers: `ignoreCase` and `treatMissingAsEmpty`
```

File: site/content/docs/main/config/request-routing.md [hunk 7]
```diff
@@ -117,6 +122,10 @@ For `header` conditions there is one required field, `name`, six operator fields
 
 - `regex` is a string representing a regular expression, and checks that the header value matches against the given regular expression.
 
+Modifiers:
+- `ignoreCase`: IgnoreCase specifies that string matching should be case insensitive. It has no effect on the `Regex` parameter.
+- `TreatMissingAsEmpty`: specifies if the header match rule specified header does not exist, this header value will be treated as empty. Defaults to false. Unlike the underlying Envoy implementation this is **only** supported for negative matches (e.g. NotContains, NotExact).
```
Comment (skriss) [inline site/content/docs/main/config/request-routing.md:127]:
```suggestion
- `treatMissingAsEmpty`: specifies if the header match rule specified header does not exist, this header value will be treated as empty. Defaults to false. Unlike the underlying Envoy implementation this is **only** supported for negative matches (e.g. NotContains, NotExact).
```

**Review by skriss — CHANGES_REQUESTED** [review]
Couple tiny nits.

Re: only implementing for negative match rules, that seems reasonable to me since those are the cases where the current behavior is somewhat unintuitive/users may want to change it. It doesn't seem to add any value for the positive cases.

File: apis/projectcontour/v1/httpproxy.go (hunk 5 above)
Comment (davinci26) [reply apis/projectcontour/v1/httpproxy.go]: fixed in bdd8b20a58135f00906b651196b6240e56dba593

File: site/content/docs/main/config/request-routing.md (hunk 6 above)
Comment (davinci26) [reply site/content/docs/main/config/request-routing.md]: fixed in bdd8b20a58135f00906b651196b6240e56dba593

File: site/content/docs/main/config/request-routing.md (hunk 7 above)
Comment (davinci26) [reply site/content/docs/main/config/request-routing.md]: fixed in bdd8b20a58135f00906b651196b6240e56dba593

**Comment by davinci26** [comment]
first 🟢  in a while

**Review by skriss — APPROVED** [review]

**Review by clayton-gonsalves — APPROVED** [review]

**Pull request merged**